vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,907 @@
+}
+
+void print_decision_scores(int f, decision_scores_t& decision_scores)
+{
+  if (f >= 0)
+  {
+    std::stringstream ss;
+    for (auto slot : decision_scores)
+    {
+      std::string delimiter = "";
+      for (auto action_score : slot)
+      {
+        ss << delimiter << action_score.action << ':' << action_score.score;
+        delimiter = ",";
+      }
+      ss << '\n';
+    }
+    ss << '\n';
+    ssize_t len = ss.str().size();
+    ssize_t t = io_buf::write_file_or_socket(f, ss.str().c_str(), (unsigned int)len);
+    if (t != len)
+      std::cerr << "write error: " << strerror(errno) << std::endl;
+  }
+}
+
+void print_update(vw& all, std::vector<example*>& slots, decision_scores_t& decision_scores, size_t num_features)
+{
+  if (all.sd->weighted_examples() >= all.sd->dump_interval && !all.quiet && !all.bfgs)
+  {
+    std::string label_str = "";
+    std::string delim = "";
+    int counter = 0;
+    for (auto slot : slots)
+    {
+      counter++;
+
+      auto outcome = slot->l.conditional_contextual_bandit.outcome;
+      if (outcome == nullptr)
+      {
+        label_str += delim;
+        label_str += "?";
+      }
+      else
+      {
+        label_str += delim;
+        label_str += std::to_string(outcome->probabilities[0].action);
+        label_str += ":";
+        label_str += std::to_string(outcome->cost);
+      }
+
+      delim = ",";
+
+      // Stop after 2...
+      if (counter > 1 && slots.size() > 2)
+      {
+        label_str += delim;
+        label_str += "...";
+        break;
+      }
+    }
+    std::ostringstream label_buf;
+    label_buf << std::setw(all.sd->col_current_label) << std::right << std::setfill(' ') << label_str;
+
+    std::string pred_str = "";
+    delim = "";
+    counter = 0;
+    for (auto slot : decision_scores)
+    {
+      counter++;
+      pred_str += delim;
+      pred_str += std::to_string(slot[0].action);
+      delim = ",";
+
+      // Stop after 3...
+      if (counter > 2)
+      {
+        pred_str += delim;
+        pred_str += "...";
+        break;
+      }
+    }
+    std::ostringstream pred_buf;
+    pred_buf << std::setw(all.sd->col_current_predict) << std::right << std::setfill(' ') << pred_str;
+
+    all.sd->print_update(all.holdout_set_off, all.current_pass, label_buf.str(), pred_buf.str(), num_features,
+        all.progress_add, all.progress_arg);
+  }
+}
+
+void output_example(vw& all, ccb& /*c*/, multi_ex& ec_seq)
+{
+  if (ec_seq.size() <= 0)
+    return;
+
+  std::vector<example*> slots;
+  size_t num_features = 0;
+  float loss = 0.;
+
+  // Should this be done for shared, action and slot?
+  for (auto ec : ec_seq)
+  {
+    num_features += ec->num_features;
+
+    if (ec->l.conditional_contextual_bandit.type == CCB::example_type::slot)
+    {
+      slots.push_back(ec);
+    }
+  }
+
+  // What does it mean for not all of the slots to be labeled? Does it become a non-labeled example at that point?
+  // Is it hold out?
```

Comment:
Having only a subset of examples labeled seems like a malformed example.  Maybe test and warn?

Reply:
Okay, can warn about this